### ISSUE TITLE
508-defect-2: Focus should go to the interactive element within a container, not the container itself, in situations where an interactive element is visible within a container that has visible focus

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#53986]
## Point of contact

joshkimux

## Severity level

2, Serious. Should be fixed in 1-2 sprints post-launch.

## Details

https://user-images.githubusercontent.com/14154792/220473394-9123e5ba-0233-4d28-8ac9-58025a68e275.mov

<br/>

> In this video, I'm using the `space` key to attempt to check this checkbox. I'm unable to do so as the focus is on the container, not the checkbox itself.

If I remember correctly, this is a bit of a wicked issue with how our components are set up where we can't focus within a component (e.g. specifically target the checkbox).

This also returns to the debate around showing visible focus or not showing visible focus on non-interactive elements.

[According to research](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/), showing visible focus is helpful to users with low vision or with cognitive disabilities to track where they are on a page or what they should literally be focusing on. 
- On the flip side, it's a false affordance as it implies that whatever has a visible focus is also interactive.

For example, take these two scenarios:
- If we show visible focus on a heading with no other signifiers that it is interactive (e.g. no underline, color, etc), users may be less confused if they attempt to use `space` or `enter` to interact with it.
- However, in this scenario, if we were to show visible focus around a checkbox which is by all means interactive, users may become very confused if they cannot use `space` or `enter` to interact with it despite the visible focus being shown.

## Reproduction steps

1. [Follow login instructions here](https://github.com/department-of-veterans-affairs/va.gov-team/issues/52046)
2. On the new and relevant evidence page, click continue without checking the checkbox. 
3. Confirm that a visible focus halo encapsulates the checkbox region. 
4. Use the space key. 
5. Confirm that you cannot deselect the checkbox.
6. On the contestable issues page, change a user entered issue. 
7. Click cancel. 
8. Confirm that a visible focus halo encapsulates the checkbox region. 
9. Use the space key. 
10. Confirm that you cannot deselect the checkbox.

## Proposed solution or next steps

**We may be able to explore the following options:**
- Hide ALL visible focus for anything that is not interactive (including headings). This is, in theory, the easiest solution as it is a consistent blanket fix. It does come at the cost of making literal attentive focus harder for folk with cognitive disabilities and low vision.
- When returning focus to anything encapsulating an interactive element, send focus directly to that interactive element and not its container. This comes at the pro of being able to keep the benefits of visible focus at the cost of a harder implementation path. We'd likely need to consider all the possible variations that might emerge, and work within the potential limitations of design system components (not sure if we can/cannot send focus onto elements and not their containers).

Before doing so, we should:
- [x] Bring this to platform office hours to determine its severity and align on recommendation
- [ ] Determine if this is feasible with the design system

### Platform recommendation
- Send focus to the interactive element within a container, not the container itself, in situations where an interactive element is visible within a container that has visible focus.

### Success criteria
If this is fixable/feasible within the design system, we should be fix this in the following cases through the reproduction steps above:
- [x] Contestable issues checkbox
- [x] New and relevant issues checkbox
- [x] Private medical records
- [x] Notice of evidence needed 
- [ ] Edit facility @rocio-desantiago this is the last one I think!

## References, articles, or WCAG support

1. [What we learned from user testing of accessible client-side routing techniques with Fable Tech Labs](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/)

## Type of issue

- [ ] Screenreader
- [ ] Keyboard
- [X] Focus
- [ ] Headings
- [ ] Color
- [ ] Zoom
- [ ] Semantics
- [ ] Axe-core
- [ ] Something else

**Comment by joshkimux** [comment]
Platform recommends to aim for the second solution
- Focus should target the interactive element within the container
- Defect-2 feels like the appropriate defect level

@Mottie I think our forms library is now set up to let us target stuff within a container, but unsure. If we cannot, we should redirect to design systems team and list as blocked.

**Comment by joshkimux** [comment]
@saderagsdale updated this with platform recommendation!

**Comment by saderagsdale** [comment]
@eileen-coforma @Mottie @rocio-desantiago hey all - does the platform's recommendations track with what you think is the best approach? Any concerns about complexity?

**Comment by Mottie** [comment]
I think the fix will be fairly straight-forward. Rocio knows what the work entails and I think 1-2 story points would be adequate.

**Comment by rocio-desantiago** [comment]
Paths:
[/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/contestable-issues](https://staging.va.gov/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/contestable-issues)
[/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/opt-in](https://staging.va.gov/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/opt-in)
[/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/notice-of-evidence-needed](https://staging.va.gov/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/notice-of-evidence-needed)
[/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/supporting-evidence/private-medical-records-authorization#privacy-agreement](https://staging.va.gov/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/supporting-evidence/private-medical-records-authorization#privacy-agreement)

**Comment by rocio-desantiago** [comment]
For this one: [/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/contestable-issues](https://staging.va.gov/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/contestable-issues)

When click on `edit`, if nothing changes, focus goes back to `edit` link. 
If `Update` pressed, focus goes back to focus to read out loud the update - which for screenreaders helps the user hear that the update was successful.

**Comment by joshkimux** [comment]
### Contestable Issues 🙆‍♂️ 

https://user-images.githubusercontent.com/14154792/224695785-0a2c6e55-f566-44c2-b7ad-8e78a0234a63.mov


### Notice of Evidence needed 🙆‍♂️ 

https://user-images.githubusercontent.com/14154792/224695599-3a0163f5-9402-4b35-9863-9e10a99dee0c.mov

### Private Medical Records 🙆‍♂️ 

https://user-images.githubusercontent.com/14154792/224696332-7717fdf3-666d-4875-a3b0-ddecc5040429.mov

### Edit facility 🙅 

@rocio-desantiago I found another one 😭 
<img width="744" alt="Screenshot 2023-03-13 at 8 01 40 AM" src="https://user-images.githubusercontent.com/14154792/224696097-1d0dab99-e139-4c74-b157-65f6a248c535.png">

**Comment by Mottie** [comment]
@joshkimux For the unchecked group error, do we focus on the first checkbox (only one showing in this scenario, but there could be more)? It's focusing on the checkbox group so that the error is read out.

**Comment by joshkimux** [comment]
@Mottie that's such a good question, and I completely missed that this can have multiple conditions listed, in which case this becomes a bit awkward as a pattern (otherwise we'd have to change how all checkboxes operate in terms of focus 😓 )

@rocio-desantiago scratch what I said earlier, this isn't an issue we should fix given that (a) there can be multiple checkboxes and (b) the checkbox group is being focused. This is likely the odd exception given the circumstances.

We can close this out 😄
